Add instructions on how to use multiple frameworks at the same time in TypeScript (#1130)
withastro/docs: src/pages/en/guides/typescript.md

}
```

If your project uses a [UI framework](/en/core-concepts/framework-components/), additional settings depending on the framework might be needed. Please see your framework's TypeScript documentation for more information. ([Vue](https://vuejs.org/guide/typescript/overview.html#using-vue-with-typescript), [React](https://reactjs.org/docs/static-type-checking.html), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript))

## Type Imports

Use explicit type imports and exports whenever possible. 

```diff
- import { SomeType } from './script';
+ import type { SomeType } from './script';
```
This way, you avoid edge cases where Astro's bundler may try to incorrectly bundle your imported types as if they were JavaScript.

In your `.tsconfig` file, you can instruct TypeScript to help with this. The [`importsNotUsedAsValues` setting](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues) can be set to `error`. Then, TypeScript will check your imports and tell you when  `import type` should be used.

> ```json
> // tsconfig.json
> {
>   "compilerOptions": {
>     "importsNotUsedAsValues": "error",
>     }
>   }
> ```

Reviewer: ``````suggestion
```json
// tsconfig.json
{
  "compilerOptions": {
    "importsNotUsedAsValues": "error",
   }
 }
```
``````

}
```

If your project uses a [UI framework](/en/core-concepts/framework-components/), additional settings depending on the framework might be needed. Please see your framework's TypeScript documentation for more information. ([Vue](https://vuejs.org/guide/typescript/overview.html#using-vue-with-typescript), [React](https://reactjs.org/docs/static-type-checking.html), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript))

## Type Imports

## Type Imports

Use explicit type imports and exports whenever possible. 

```diff
- import { SomeType } from './script';
+ import type { SomeType } from './script';
```
This way, you avoid edge cases where Astro's bundler may try to incorrectly bundle your imported types as if they were JavaScript.

In your `.tsconfig` file, you can instruct TypeScript to help with this. The [`importsNotUsedAsValues` setting](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues) can be set to `error`. Then, TypeScript will check your imports and tell you when  `import type` should be used.

> ```json
> {
>   "compilerOptions": {
>     "importsNotUsedAsValues": "error",

Reviewer: ```suggestion
```json
// tsconfig.json
{
  "compilerOptions": {
    "importsNotUsedAsValues": "error",
    }
  }
```
I think I want the parentheses closed @Jutanium ! We do so earlier, showing only the exact config in question, not the entire thing, so I'm going to run with this. We will trust people will get this, until we have evidence to the contrary.

}
```

If your project uses a [UI framework](/en/core-concepts/framework-components/), additional settings depending on the framework might be needed. Please see your framework's TypeScript documentation for more information. ([Vue](https://vuejs.org/guide/typescript/overview.html#using-vue-with-typescript), [React](https://reactjs.org/docs/static-type-checking.html), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript))

> ## Type Imports
> 
> 
> 

Reviewer: ```suggestion
```
We ended up with this heading twice.

}
```

> If your project uses a framework, additional settings depending on the framework might be needed, see your framework's documentation for more information. ([Vue](https://vuejs.org/guide/typescript/overview.html#using-vue-with-typescript), [React](https://reactjs.org/docs/static-type-checking.html), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript))

Reviewer: ```suggestion
If your project uses a [UI framework](/en/core-concepts/framework-components/), additional settings depending on the framework might be needed. Please see your framework's TypeScript documentation for more information. ([Vue](https://vuejs.org/guide/typescript/overview.html#using-vue-with-typescript), [React](https://reactjs.org/docs/static-type-checking.html), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript))
```

:::

📚 Read more about [`.ts` file imports](/en/guides/imports/#typescript) in Astro.
📚 Read more about [TypeScript Configuration](https://www.typescriptlang.org/tsconfig/).

## Troubleshooting

### Errors Typing multiple JSX frameworks at the same time

An issue may arise when using multiple JSX frameworks in the same project, as each framework requires different, sometimes conflicting, settings inside `tsconfig.json`.

**Solution**: Set the [`jsxImportSource` setting](https://www.typescriptlang.org/tsconfig#jsxImportSource) to `react` (default), `preact` or `solid-js` depending on your most-used framework. Then, use a [pragma comment](https://www.typescriptlang.org/docs/handbook/jsx.html#configuring-jsx) inside any conflicting file from a different framework.

For the default setting of `jsxImportSource: react`, you would use:

```jsx
// For Preact
/** @jsxImportSource preact */

// For Solid
/** @jsxImportSource solid-js */
```

### Vue components are mistakenly typed by the `@types/react` package when installed

> The types definition from the `@types/react` package are declared globally and therefore will be mistakenly used to typecheck `.vue` files when using [Volar](https://github.com/johnsoncodehk/volar).

Reviewer: ```suggestion
The types definitions from the `@types/react` package are declared globally and therefore will be mistakenly used to typecheck `.vue` files when using [Volar](https://github.com/johnsoncodehk/volar).
```

📚 Read more about [`.ts` file imports](/en/guides/imports/#typescript) in Astro.  
📚 Read more about [TypeScript Configuration](https://www.typescriptlang.org/tsconfig/).

## Troubleshooting

> ### Typing multiple JSX frameworks at the same time

Reviewer: ```suggestion
### Errors Typing multiple JSX frameworks at the same time
```
Making this read more like a problem to fix, so that we can make it a similar form to our Troubleshooting page

> An issue arise when using multiple JSX frameworks in the same project, as every framework require different, conflicting settings inside `tsconfig.json`. To workaround this issue, TypeScript support [different pragma comments](https://www.typescriptlang.org/docs/handbook/jsx.html#configuring-jsx) to set the JSX options on a per-file basis.

Reviewer: ```suggestion
An issue may arise when using multiple JSX frameworks in the same project, as each framework requires different, sometimes conflicting, settings inside `tsconfig.json`. 

**Solution**: Set the [`jsxImportSource` setting](https://www.typescriptlang.org/tsconfig#jsxImportSource) to `react` (default), `preact` or `solid-js` depending on your most-used framework. Then, use a [pragma comment](https://www.typescriptlang.org/docs/handbook/jsx.html#configuring-jsx) inside any conflicting file from a different framework.

For the default setting of 	`jsxImportSource: react`, you would use:
```

Author: Ah! Your wording is so much better here, I rewrote that part like 5 times haha

> The [`jsxImportSource` setting](https://www.typescriptlang.org/tsconfig#jsxImportSource) being by default `react`, all your `.jsx` files will by default be typechecked using the React definitions if installed (`@types/react`). As such, your Preact and Solid files should contain pragma comments to set the setting to the proper value:

Reviewer: ```suggestion
```

```jsx
// For Preact
/** @jsxImportSource preact */

// For Solid
/** @jsxImportSource solid-js */
```

> Alternatively, if your project primarily use another framework, you may change the `jsxImportSource` setting to `preact` or `solid-js` and instead use a pragma comment inside your React files.

Reviewer: ```suggestion
```

> ### Typing Vue components when the `@types/react` package is installed
> 
> Due to being declared globally, the types definition from the `@types/react` package, if installed, will mistakenly be used to typecheck `.vue` files when using [Volar](https://github.com/johnsoncodehk/volar). There's currently no reliable way to fix this, however a few solutions are discussed can be found in [this GitHub discussion](https://github.com/johnsoncodehk/volar/discussions/592)

Reviewer: ```suggestion
### Vue components are mistakenly typed by the `@types/react` package when installed

The types definition from the`@types/react` package is declared globally and therefore will be mistakenly used to typecheck `.vue` files when using [Volar](https://github.com/johnsoncodehk/volar).

**Status**:  Expected behavior.

**Solution**: There's currently no reliable way to fix this, however a few solutions and more discussion can be found in [this GitHub discussion](https://github.com/johnsoncodehk/volar/discussions/592).

```
Attempting to make this look more like the entries from our TroubleShooting page, emphasizing the status of the behaviour and solution.
